zkevm-circuits/src/circuit_tools/cell_manager.rs

```diff
@@ -284,6 +284,13 @@ impl<F: Field, C: CellType> CellManager<F, C> {
         }
     }
 
+    pub(crate) fn restart(&mut self) {
+        self.height = self.height_limit;
+        for col in self.columns.iter_mut() {
+            col.height = 0;
+        }
+    }
```

Comment:
Used in PI refactoring

Reply:
So in current PI constraints are in different gates:
```
let cb = new ConstraintBuilder(..);
meta.create_gate("gate_1", |meta| { 
    circuit!([meta, cb] {
            //... 
    })
};
meta.create_gate("gate_2", |meta| { 
    cb.restart();
    circuit!([meta, cb] {
            //... 
    })
};
```
In MPT we don't have that concerns at all cuz there's only one big gate, for separate gate tho `reset` in CM only is not enough. I think we should keep it cuz it allow ppl to still use multiple gates with one CB.

Reply:
People shouldn't use different gates with the circuit tools. So just do the same thing for the PI circuit as the MPT circuit does, just move everything to a single gate which is much cleaner. (I don't like the gate API of halo2).

Reply:
In the future even that single gate will be removed, now it's just necessary to get the necessary halo2 object returned in the gate. In the future even that single gate itself should be hidden from the user.

Reply:
> In the future even that single gate will be removed, now it's just necessary to get the necessary halo2 object returned in the gate. In the future even that single gate itself should be hidden from the user.
😯 Is this your plan with the Tiquito DSL? Pretty wild haha. But I also know some Halo2 user who like the gate system and actually implement their `Layouter` (the [ZKML project ](https://github.com/zkonduit/ezkl), I talked to Dante who's the core dev and he said that it's actually powerful), so idk if we should leave it as a option before going too far?

Reply:
Was always the plan, but halo2 didn't allow me to do it!

I don't see the point of the gates API of halo2. If you use a single gate or multiple gates, internally for halo2 it doesn't matter. It also forces you to divide your constraints up in multiple sections which is pretty strange, better to just use things people are already familiar with like functions to divide their code in parts that make sense.

I'm also not a fan of the layouter, but I don't see how that's directly related to the gates API.

Reply:
I guess Gates define a set of constraints that has to be together in fixed relative positions, and Layouter measures the max height and width of gates and rearrange them, it's useful when you don't have many inter-depending logic. For example in ZKML, one gate could be one tensor, you operate independently on tensor but they can be shift around with your implemented [layouter](https://github.com/zkonduit/ezkl/blob/7ffd6460277c24cde87bac534a750e4a1b628670/src/circuit/modules/planner.rs#L23).

Reply:
As far as I know the layouter only takes into account the region as defined in synthesize, because on the constraint level halo2 has no idea on what rows the constraints will actually be active (the same constraint may need to be active on many rows in a chip). So the height is calculated depending on the number of rows filled in in synthesize.